File: service/internal/infrastructure/health/service.go
package health

import (
	"context"

	"github.com/rs/zerolog/log"
)

// Service is the standard implementation of the healthchecker.
type Service struct {
	Components map[string]Healthchecker
}

// CheckHealth will check the overall system health and return this.
func (s Service) CheckHealth(ctx context.Context) SystemHealth {
	components := map[string]ComponentHealth{}
	for name, component := range s.Components {
		componentHealth := component.Healthcheck(ctx)
		log.Debug().
			AnErr("err", componentHealth).
			Bool("healthy", componentHealth == nil).
			Str("component", name).
			Msg("Health of component")

		components[name] = ComponentHealth{componentHealth}
	}

	return SystemHealth{components}
}

File: service/health/src/service.rs
use super::{model::ComponentHealth, model::SystemHealth, Healthchecker};
use std::{collections::HashMap, sync::Arc};

/// Service layer for checking the health of the system.
#[derive(Clone)]
pub struct HealthService {
    /// The map of components to check the health of.
    components: HashMap<String, Arc<dyn Healthchecker>>,
}

impl HealthService {
    /// Create a new instance of the health service.
    ///
    /// # Parameters
    /// - `components` - The map of components to check the health of.
    ///
    /// # Returns
    /// The health service
    #[must_use]
    pub fn new(components: HashMap<String, Arc<dyn Healthchecker>>) -> Self {
        Self { components }
    }

    /// Actually check the health of the system.
    ///
    /// # Returns
    /// The overall system health
    #[tracing::instrument(skip(self))]
    pub async fn check_health(&self) -> SystemHealth {
        tracing::debug!("Checking system health");

        let mut components = HashMap::new();
        for (name, component) in &self.components {
            let result = component.check_health().await;

            let component_health = match result {
                Ok(_) => ComponentHealth::Healthy,
                Err(e) => ComponentHealth::Unhealthy(e.to_string()),
            };

            tracing::debug!(name = ?name, status = ?component_health, "Component health");
            components.insert(name.clone(), component_health);
        }

        SystemHealth { components }
    }
}
